### ISSUE TITLE
Muted Autoplay not working in Chrome 66

### ISSUE THREAD

**Issue by codener** [flowplayer/flowplayer#1379]
I have an issue where Flowplayer refuses to use muted autoplay. Digging through the code, I have found the culprit:

https://github.com/flowplayer/flowplayer/blob/bad434c8801be8f8523b0ef3a0545133303593f7/lib/engine/html5-factory.js#L121-L138

The code tries to autoplay first. If that fails, it mutes the player and retries autoplaying. While `play()` is called right after the `mute()` in line 130, the issue appears to be that the `mute()` function is *async*. 

https://github.com/flowplayer/flowplayer/blob/bad434c8801be8f8523b0ef3a0545133303593f7/lib/engine/html5-factory.js#L164-L168

I haven't been able to reproduce this outside of my test environment, but I believe the issue is clear. I will append to this when I know more.

In meantime, could you have a look? To me, this seems quite obvious.

---
This might be related to #1378.

**Comment by codener** [comment]
As far as I can say by now, the issue is not `mute()` being async or so.

However, I'm getting a `DOMException: The play() request was interrupted by a new load request` on the `play()` in line 131. I have no clue on why yet.

**Comment by codener** [comment]
I blame the `api.play()` in line 131 for the error – but I have no idea what's wrong. Although it is invoked from the promise rejection, the `api` (aka video element) doesn't seem to be ready for another `play()` invocation.

I should mention that about 1 in 10 times, it works as expected. So there is a non-deterministic aspect :disappointed: 

Working around by injecting this code before the first `play()` in line 121:
```
          if (conf.mutedAutoplay)
            player.mute(true, true);
```

**Comment by danrossi** [comment]
Hi this is working. I am also trying to get some work arounds in my streaming features.  However you only need to mute on first playback. 

The browser determines and dictates this. Perhaps there is some bad ratio thing. But on second visit on that domain playback can begin unmuted. 

I don't think the autoplay work around is controlled on the api but for the html5 engine. But this is sadly the basis of the work around to detect further page loads and not configure to mute. It prob should be integrated into the part that controls the sources and engine selector. 

```
 if (this.autoPlay) {
            if (!window.localStorage.visited) {
                window.localStorage.visited = true;
                this.startMuted = true;

            }
        }
```

**Comment by ondreian** [comment]
The ability to autoplay in Google Chrome 66+ is driven by MEI (Media Engagement Index) and this is determined by much more than just visiting a website.  

1. Size of the `HTMLVideoElement` that is playing the video (if it's a small video it does not increase MEI)
2. amount of time spent watching a video
3. the ratio of visits versus Media element engagements
4. Several other factors play a role in determining if autoplay is successful

As the algorithm for determining MEI is refined it will only become more complex and the `visited` solution more brittle.

> Here's the one thing to remember: Don't ever assume a video will play, and don't show a pause button when the video is not actually playing. It is so important that I'm going to write it one more time below for those who simply skim through that post.

This quote is taken directly from the [Google Autoplay Policy updates](https://developers.google.com/web/updates/2017/09/autoplay-policy-changes)

**Comment by codener** [comment]
@danrossi I hardly understand your message, I'm sorry. I believe muting the player before the first `play()` when "mute" and "mutedAutoplay" are configured/requested is always right, and might even fulfill the expectations of a user better than the current behaviour. Still, the current implementation *should* be working.

@ondreian Thanks for your reply. I should've mentioned that I'm aware of the whole MEI/muted autoplay/no autoplay with sound issue. The Flowplayer code is correct in terms of your Google documentation quote, but unfortunately it's broken in terms of reality :disappointed: 

Apparently I need to debug further. When I use a fresh Chrome instance with a fresh user data dir, muted autoplay is working. Maybe it's something in my test setup, but still strange.

**Comment by nnarhinen** [comment]
@codener and your problems exist with FP 7.2.6? Could you point to a setup where this does happen for you? In my local tests I don't have this problem with any chrome versions tested. Of course we will fix this if we manage to get a reproducable test case.

**Comment by codener** [comment]
I can confirm the issue persists with 7.2.6. Will provide a reproduction if I can.

**Comment by strarsis** [comment]
+1! The video doesn't require audio so it would be great if it is autoplayed.

https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
> Muted autoplay is always allowed.

**Comment by nnarhinen** [comment]
We still need an example setup and instructions on how to reproduce this consistently to fix this.

**Comment by codener** [comment]
Updating HLS.js from 0.8.1 to 0.9.1 seems to have fixed most issues, including missing/incomplete error handling.
I consider this issue resolved. Will reopen if trouble returns. Thanks!

**Comment by strarsis** [comment]
@codener: So I just have to update flowplayer in order to also update HLS.js for fixing this issue?

**Comment by codener** [comment]
@strarsis I believe your issue is different from mine, as I was experiencing a real edgy edge case.

I think the best would be to [open a new issue](/flowplayer/flowplayer/issues/new) and describe your issue as detailed as possible, following the [contribution guide](https://github.com/flowplayer/flowplayer/blob/dev/CONTRIBUTING.md).

**Comment by codener** [comment]
As a quick follow-up: I could not apply the upgrade (FP + HLS) because of `DOMException`s thrown during stream (media source) changes. I'm sticking with the aforementioned workaround (see my 3rd post).

**Comment by eligence** [comment]
We had this same issue and discovered that the flowplayer documentation has been updated to specify that hls scripts mus be loaded before the core flowplayer script. In previous version the opposite was true. Additionally, in the hls demos hls is loaded after flowplay js. This is incorrect.

See updated hls setup here. I have tested this and it works with version 7.2.7
https://flowplayer.com/help/developers/flowplayer-7/setup#hls
